pkg/docker/config/config.go

```diff
@@ -37,111 +75,247 @@ var (
+	if credHelper, exists := s.Config.CredHelpers[serverURL]; exists {
+		helperName := fmt.Sprintf("docker-credential-%s", credHelper)
+		p := helperclient.NewShellProgramFunc(helperName)
+		creds, err := helperclient.Get(p, serverURL)
+		if err != nil {
+			return "", "", err
+		}
+		return creds.Username, creds.Secret, nil
+	}
+
+	// I'm feeling lucky
+	if val, exists := s.Config.Auths[serverURL]; exists {
+		return val.Username, val.Secret, nil
+	}
+
+	// bad luck; let's normalize the entries first
+	serverURL = normalizeRegistry(serverURL)
+	normalizedAuths := map[string]Auth{}
+	for k, v := range s.Config.Auths {
+		normalizedAuths[normalizeRegistry(k)] = v
+	}
+	if val, exists := normalizedAuths[serverURL]; exists {
+		return val.Username, val.Secret, nil
+	}
+	return "", "", credentials.NewErrCredentialsNotFound()
+}
+
+// Delete removes credentials from the store.
+func (s *AuthStore) Delete(serverURL string) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.Config == nil {
+		err := s.load()
+		if err != nil {
+			if credentials.IsErrCredentialsNotFound(err) {
+				return nil
+			}
+			return err
+		}
+	}
+
+	// First try cred helpers.
+	if credHelper, exists := s.Config.CredHelpers[serverURL]; exists {
+		helperName := fmt.Sprintf("docker-credential-%s", credHelper)
+		p := helperclient.NewShellProgramFunc(helperName)
+		return helperclient.Erase(p, serverURL)
+	}
+
+	changed := false
+	if _, ok := s.Config.Auths[serverURL]; ok {
+		delete(s.Config.Auths, serverURL)
+		changed = true
+	} else if _, ok := s.Config.Auths[normalizeRegistry(serverURL)]; ok {
+		delete(s.Config.Auths, normalizeRegistry(serverURL))
+		changed = true
+	}
+	if changed {
+		return s.save()
+	}
+
+	return nil
+}
+
+// SetAuthentication stores the username and password in the auth.json file
+//
+// Deprecated: Use an AuthStore.
+func SetAuthentication(sys *types.SystemContext, registry, username, password string) error {
+	authStore := &AuthStore{Path: sys.AuthFilePath}
+	return authStore.Add(&credentials.Credentials{
+		ServerURL: registry,
+		Username:  username,
+		Secret:    password,
 	})
 }
 
 // GetAuthentication returns the registry credentials stored in
 // either auth.json file or .docker/config.json
 // If an entry is not found empty strings are returned for the username and password
+//
+// Deprecated: Use an AuthStore.
 func GetAuthentication(sys *types.SystemContext, registry string) (string, string, error) {
-	if sys != nil && sys.DockerAuthConfig != nil {
-		return sys.DockerAuthConfig.Username, sys.DockerAuthConfig.Password, nil
+	authStore := &AuthStore{}
+
+	if sys != nil {
+		if sys.DockerAuthConfig != nil {
+			return sys.DockerAuthConfig.Username, sys.DockerAuthConfig.Password, nil
+		}
+
+		if sys.AuthFilePath != "" {
+			authStore.Path = sys.AuthFilePath
+		} else if sys.RootForImplicitAbsolutePaths != "" {
+			authStore.Path = filepath.Join(sys.RootForImplicitAbsolutePaths, fmt.Sprintf(defaultPerUIDPathFormat, os.Getuid()))
+		}
```

Comment:
Why was this moved out of `getPathToAuth`/`getPath`? Now only `GetAuthentication`, but not `SetAuthentication`/`RemoveAuthentication` implement the `SystemContext` options.

Reply:
> Why was this moved out of `getPathToAuth`/`getPath`?

Because I want the `AuthStore` struct to be consumed by `SystemContext`, and don't want a dependency cycle.

> Now only `GetAuthentication`, but not `SetAuthentication`/`RemoveAuthentication` implement the `SystemContext` options.

Fixed with 8ce2b8b -> 5e011a3.